client/main/contextmenu.js

```diff
@@ -0,0 +1,97 @@
+
+'use strict';
+
+const $ = require('jquery');
+const Backbone = require('backbone');
+
+const tarp = require('./utils/tarp');
+
+const ContextMenu = function() {
+
+    Object.assign(this, Backbone.Events);
+
+    this.$el = $('<div class="jmv-context-menu jmv-ribbon-group-body-vertical"></div>');
+    this.$el.hide();
+
+    this.$el.on('menuActioned', () => { this._menuClosed(); });
+
+    this.owner = null;
+
+    this.show = function(menuItems, x, y, openPath, owner) {
+        openPath = openPath === undefined ? [] : openPath;
+        this.owner = owner;
+        this.$el.show();
+        if ( ! this._tarpVisible) {
+            tarp.show('click-menu', true, 0, 40)
+                .then(() => this._menuClosed(), (event) => this._menuClosed(event));
+            this.$el.css('z-index', '50');
+            this._tarpVisible = true;
+        }
+
+        this.buttons = [ ];
+        let menuShown = (menu) => this._menuShown(menu);
+
+        this.$el.empty();
+        this.$separator = $('<div class="jmv-click-menu-separator"></div>').appendTo(this.$el);
+
+        let openButton = null;
+
+        for (let i = 0; i < menuItems.length; i++) {
+            let button = menuItems[i];
+
+            if (button.dock === 'right')
+                button.$el.insertAfter(this.$separator);
+            else
+                button.$el.insertBefore(this.$separator);
+            button.on('shown', menuShown);
+            this.buttons.push(button);
+
+            if (openPath.length > 0 && button.openPath) {
+                openButton = button.openPath(openPath);
+                if (openButton !== null)
+                    openPath = [];
+            }
+        }
+
+        if (openButton !== null) {
+            x -= this.$el.outerWidth(true) - 10;
+            y -= openButton.$el.position().top + 10;
+        }
+
+        this.$el.css({ top: y, left: x });
+
+    };
+
+    this.isVisible = function() {
+        return this._tarpVisible;
```

Comment:
should this just be `_visible`, or `_isVisible`?

Reply:
It is used outside this object, so technically no underscore.

Reply:
i mean `this._tarpVisible` should probably just be `this._isVisible` ... that the tarp is visible isn't as significant as the whole thing being visible.

Reply:
sorry are you talking about line 66 or 65?

Reply:
66

Reply:
in which case yep